enhancements/agent-installer/automated-workflow.md

```diff
@@ -0,0 +1,491 @@
+   installer.
+5. The orchestration tool prepares an `install-config.yaml` and
+   separate manifest files with `NMState` content using input data and
+   templates.
+6. The orchestration tool runs `openshift-install create image`.
+7. The installer extracts a copy of the RHCOS ISO from the OpenShift
+   release payload.
+8. The installer combines the inputs provided by the user and produces
+   an Ignition config.
+9. The installer combines the Ignition config and the RHCOS ISO to
+   create the bootable image configured to run the assisted service or
+   agent based on the host that boots the image and writes the new
+   image and cluster credentials to disk.
+10. The orchestration tool copies the ISO to a location that the hosts
+    or VMs can boot from it. For bare metal, that will generally be a
+    web server visible to the management controllers in the bare metal
+    hosts (because the management controllers are unlikely to be on
+    the same network as the external interfaces for the cluster
+    hosts). For vSphere or other hypervisors, that may be a filesystem
+    on the hypervisor host.
+11. The orchestration tool configures the host to boot the ISO. For
+    bare metal, that means configuring the baseboard management
+    controller (BMC) in each host to expose the ISO through the
+    virtual media interface. For vSphere or other hypervisors, similar
+    settings of the VM need to be adjusted.
+12. The orchestration tool boots the host.
+13. The orchestration tool uses `openshift-install wait-for
+    install-complete` to watch the deployment progress and wait for it
+    to complete.
+14. On each host, the startup script in the image selects the correct
+    network settings based on the MAC addresses visible and applies
+    the network settings for each interface.
+15. On node0, the startup script in the image recognizes that it is on
+    node0 and launches the assisted service and the create-cluster
+    service to drive the deployment.
+16. On node0, the create-cluster service waits for the assisted
+    service to start, then creates an InfraEnv using the API.
+17. On node0, the create-cluster service waits for all of the known
+    agents to be registered with the service.
+18. On all nodes (including node0), the startup script in the image
+    connects to the assisted service on node0 to find the InfraEnv ID
+    for the cluster being deployed (there will only be 1). After it
+    has the ID, the startup script launches the assisted agent with
+    that ID and configured to connect to the assisted service on
+    node0.
+19. All of the agents register their hosts with the assisted service.
+20. The assisted service performs its standard validations for all of the
+    hosts.
+21. On node0, the create-cluster service sees that all known agents
+    have registered and the hosts have passed validation. It uses the
+    assisted service API to trigger the deployment to proceed.
```

Comment:
@rwsu on an earlier draft you said

> Likely to be a different service, a initiate-cluster-install service. I don't know if you intended to go into this much detail, but at around the same time, something will select node0 to be the bootstrap node.

I can update this to refer to the separate service.

We should describe how node0 is selected as part of this set of steps. Is that worked out, yet?

Reply:
I think this wasn't stated explicitly, but I assume the user specifies which host will be node0 in the configuration?  So that it's clear where the REST API endpoints will be?
Also, what would this look like with DHCP?

Reply:
The agent installer does need to be able to control which node is node0. It's not entirely clear that the user cares, although they might. I had thought that we would either ask for an IP (for the DHCP case) or use the first host in the inventory (for the static networking case). But, we haven't decided which approach makes the most sense.

Reply:
> Also, what would this look like with DHCP?

There's the requirement that at least one node should have static addressing. If it's only one node, that one will be node0. There's no other discovery mechanism available.

Reply:
That requirement is a difficult one and should be made explicit

Reply:
All nodes must have IPs that don't change, even in the DHCP case. That's a fundamental cube requirement -- nodes can't change their IP.

It's not required that any nodes have static networking provided to the installer for the DHCP case, we just need to know one of the IPs to use to pick a host to be node0. That IP can be assigned by the DHCP server.